python/cudf/cudf/core/udf/function.cu

```diff
@@ -0,0 +1,555 @@
+
+  do {
+    assumed = old;
+    old =
+      atomicCAS(address_as_ull, assumed, __double_as_longlong(val + __longlong_as_double(assumed)));
+
+  } while (assumed != old);
+
+  return __longlong_as_double(old);
+}
+
+// int64_t atomicAdd
+__device__ __forceinline__ int64_t atomicAdd(int64_t* address, int64_t val)
+{
+  return atomicAdd((unsigned long long*)address, (unsigned long long)val);
+}
+
+// double atomicMax
+__device__ __forceinline__ double atomicMax(double* address, double val)
+{
+  unsigned long long old = __double_as_longlong(*address);
+  while (val > __longlong_as_double(old)) {
+    unsigned long long assumed = old;
+    if ((old = atomicCAS((unsigned long long*)address, assumed, __double_as_longlong(val))) ==
+        assumed)
+      break;
+  }
+  return __longlong_as_double(old);
+}
+
+// int64_t atomicMax
+__device__ __forceinline__ int64_t atomicMax(int64_t* address, int64_t val)
+{
+  return atomicMax((long long*)address, (long long)val);
+}
+
+// double atomicMin
+__device__ __forceinline__ double atomicMin(double* address, double val)
+{
+  unsigned long long old = __double_as_longlong(*address);
+  while (val < __longlong_as_double(old)) {
+    unsigned long long assumed = old;
+    if ((old = atomicCAS((unsigned long long*)address, assumed, __double_as_longlong(val))) ==
+        assumed)
+      break;
+  }
+  return __longlong_as_double(old);
+}
+
+// int64_t atomicMin
+__device__ __forceinline__ int64_t atomicMin(int64_t* address, int64_t val)
+{
+  return atomicMin((long long*)address, (long long)val);
+}
+
+// Use a C++ templated __device__ function to implement the body of the algorithm.
+template <typename T>
+__device__ T device_sum(T const* data, int const items_per_thread, size_type size) {
+  __shared__ T sum;
+  int tid = threadIdx.x;
+  int tb_size = blockDim.x;
+  T local_sum = 0;
+
+  if (tid == 0) sum = 0;
+
+  __syncthreads();
+
+// Calculate local sum for each thread
+#pragma unroll
+  for (size_type item = 0; item < items_per_thread; item++) {
+    if (tid + (item * tb_size) < size) {
+      T load = data[tid + item * tb_size];
+      local_sum += load;
+    }
+  }
+
+  atomicAdd(&sum, local_sum);
+
+  __syncthreads();
+
+  return sum;
+}
+
+// Use a C++ templated __device__ function to implement the body of the algorithm.
+template <typename T>
+__device__ T device_var(T const* data, int const items_per_thread, size_type size) {
+
+  int tid = threadIdx.x;
+  int tb_size = blockDim.x;
+
+  double local_var            = 0;
+  __shared__ double var;
+  if (tid == 0) var = 0;
+
+  T sum = device_sum<T>(data, items_per_thread, size);
+  double mean = sum / static_cast<double>(size);
+
+// Calculate local sum for each thread
+#pragma unroll
+  for (size_type item = 0; item < items_per_thread; item++) {
+    if (tid + (item * tb_size) < size) {
+      T load = data[tid + item * tb_size];
+      double temp  = load - mean;
+      temp         = pow(temp, 2);
+      local_var += temp;
+    }
+  }
```

Comment:
If we want a one-pass approach, might be worth considering some of the options from Chan, Golub, and LeVeque (1983), Section 4 (http://cpsc.yale.edu/sites/default/files/files/tr222.pdf). Or, when using this two-pass method, using their "corrected two pass method" (eq 1.7) for better numerical stability.